### PR TITLE
[#2462] Add flag AllowSystemComponents to worker options

### LINKED ISSUE #2462: Specify NodeSelector/Affinities for Shoot core components
**How to categorize this issue?**
/area control-plane
/kind enhancement
/priority normal

**What would you like to be added**:
I'd like to be able to specify `nodeSelector`s or `affinities` for the core component that Gardener deploys into each seed so that I can control on which worker pools those components run.

**Why is this needed**:
This is needed as some workerpools might have special hardware or might have very much resources which are reserved for workload.

By letting the core components run wherever they are scheduled they might end up on hardware which is very expensive and should only be used by the intended workload.

Allowing to specify `nodeSelector`s or `affinities` would also make it possible to create a dedicated workerpool for core components.

### PR DESCRIPTION
**Pull request by jannickfahlbusch**
**How to categorize this PR?**
/area control-plane
/kind enhancement
/kind api-change
/priority normal

**What this PR does / why we need it**:
This allows to restrict the core components from being able to be scheduled to every worker pool.

I've added the object `systemComponents` on worker pools which contains a field `allow`. This field defaults to true.

 Each node will receive a label `worker.gardener.cloud/system-components` with the specified value.

The core components received a `nodeSelector` for `worker.gardener.cloud/system-components=true`

**Which issue(s) this PR fixes**:
Fixes #2462 

**Special notes for your reviewer**:
Currently it would be possible to restrict the core-components solely to a worker-group which is scaled to zero. I think this makes no sense. What do you think?

**Release note**:
```improvement user
It is now possible to restrict core components from running on a worker pool by specifying `systemComponents.allow: false` in the pool definition.

System components deployed by extensions (Such as typha from calico) need a separate adaptation and are not covered by this change.
```

### CODE REVIEW

**Comment by jannickfahlbusch** [comment]
Somehow `make generate` does not generate the DeepCopy functions. Did I forget something?

File: pkg/apis/core/v1beta1/defaults.go
Comment (vpnachev) [inline pkg/apis/core/v1beta1/defaults.go]: Same defaulting needs to be done for v1alpha1 version, as well.

File: pkg/apis/core/validation/shoot.go
Comment (vpnachev) [inline pkg/apis/core/validation/shoot.go]: Yes, the maximum must be >=1.
As far as I know, scale-to-zero is not implemented for all providers [1], so from my point of view the check should for `minimum>=1`.

[1] https://github.com/gardener/autoscaler/issues/27

File: pkg/apis/core/validation/shoot_test.go
Comment (vpnachev) [inline pkg/apis/core/validation/shoot_test.go]: move this import in the second group of packages

**Review by vpnachev — CHANGES_REQUESTED** [review]
Thanks, looks very nice.

**Comment by vpnachev** [comment]
> Somehow `make generate` does not generate the DeepCopy functions. Did I forget something?

If you have not, run `make install-requirements` and then re-run `make generate`.

File: pkg/apis/core/validation/shoot.go
Comment (jannickfahlbusch) [reply pkg/apis/core/validation/shoot.go]: Done

**Comment by jannickfahlbusch** [comment]
Figured it out:

Since the repo is a Go module I assumed that I can place the repo outside of my `GOPATH`. When run outside of the `GOPATH`, `make generate` finishes without any error but does not generate additional sources.

**Comment by jannickfahlbusch** [comment]
At the moment the following components would respect the setting:
* CoreDNS
* Metrics server
* Blackbox exporter
* VPN

Did I miss anything?

**Comment by vpnachev** [comment]
Yeah, code generation require the repo to be checked out under $GOPATH even though we are using go modules, see https://github.com/kubernetes/kubernetes/issues/86753 for more info.

>Did I miss anything?

I think no. You would like to open a separate PR to https://github.com/gardener/gardener-extension-networking-calico/, but the change there will be incompatible with the currently released gardener versions, so I would wait or add some additional logic to set the `nodeSelector` only if gardener version support it.

/reviewed ok-to-test

**Comment by timebertt** [comment]
Thanks @jannickfahlbusch for contributing this feature, really appreciate it!


> At the moment the following components would respect the setting:
> 
> * CoreDNS
> * Metrics server
> * Blackbox exporter
> * VPN
> 
> Did I miss anything?

You can probably add it to:
- nginx-ingress-controller
- kubernetes-dashboard

/status author-action

**Comment by timebertt** [comment]
/needs review

I will try to take a look today :)
/assign

File: example/90-shoot.yaml
Comment (rfranzke) [inline example/90-shoot.yaml]: API-wise, such a structure would be more prepared in case we want to introduce further configuration settings for the system components in the future:

```suggestion
    # systemComponents:
    #   allowed: true
```

WDYT?

File: pkg/apis/core/validation/shoot.go
Comment (rfranzke) [inline pkg/apis/core/validation/shoot.go]:
```suggestion
			if worker.AllowSystemComponents != nil && *worker.AllowSystemComponents {
```

File: pkg/apis/extensions/v1alpha1/types_worker.go
Comment (rfranzke) [inline pkg/apis/extensions/v1alpha1/types_worker.go]:
```suggestion
	// AllowSystemComponents determines whether the pool should be allowed to host system components or not (defaults to true)
	// +optional
```

File: pkg/apis/extensions/v1alpha1/types_worker.go
Comment (rfranzke) [inline pkg/apis/extensions/v1alpha1/types_worker.go]:
```suggestion
	AllowSystemComponents *bool `json:"allowSystemComponents,omitempty"`
```

Though, same suggestion like above.

File: pkg/operation/botanist/worker.go
Comment (rfranzke) [inline pkg/operation/botanist/worker.go]: nil pointer check for missing

**Review by rfranzke — CHANGES_REQUESTED** [review]
Nice PR, thank you!
Can you also enhance https://github.com/gardener/gardener/blob/master/docs/extensions/worker.md with a comment about the new field that is being introduced?

File: charts/shoot-core/components/charts/coredns/templates/coredns-deployment.yaml
Comment (timuthy) [inline charts/shoot-core/components/charts/coredns/templates/coredns-deployment.yaml]:
```suggestion
        worker.gardener.cloud/system-components: true
```
"allowed" is functionally more related to taints/tolerations, i.e. if you create a system component pod w/o a label selector, nodes with `system-components.gardener.cloud/allowed: false` will also be considered.

File: pkg/apis/core/v1alpha1/constants/types_constants.go
Comment (timuthy) [inline pkg/apis/core/v1alpha1/constants/types_constants.go]:
```suggestion
	LabelWorkerPoolAllowSystemComponents = "worker.gardener.cloud/system-components"
```

**Review by timuthy — CHANGES_REQUESTED** [review]

File: pkg/apis/core/v1alpha1/defaults_test.go
Comment (timebertt) [inline pkg/apis/core/v1alpha1/defaults_test.go]:
```suggestion
			Expect(obj.AllowSystemComponents).To(PointTo(BeTrue()))
```
Same for the other API version.

File: pkg/apis/core/validation/shoot.go
Comment (timebertt) [inline pkg/apis/core/validation/shoot.go]: Hmm, should we also check, if there are no or only compatible taints on the worker pool? Otherwise we can't be sure, that the system components are indeed able to run on a worker pool.
So maybe you can just move this check down to line 1091.

File: pkg/operation/botanist/worker.go
Comment (timebertt) [inline pkg/operation/botanist/worker.go]: I think, we don't need an additional field on the Worker. The node labels are already added via `Worker.spec.pools[].labels` (as you already adapted in line 95). That way, the worker controller doesn't need to act upon this setting and doesn't even need to know about it...

**Review by timebertt — COMMENTED** [review]
I don't have to add much to the previous reviews, only some minor things.
Otherwise, looks quite good already!

File: example/90-shoot.yaml
Comment (jannickfahlbusch) [reply example/90-shoot.yaml]: Sound reasonable to me, I'll change it

File: pkg/apis/extensions/v1alpha1/types_worker.go
Comment (jannickfahlbusch) [reply pkg/apis/extensions/v1alpha1/types_worker.go]: Removed due to the comment from @tim-ebert below: https://github.com/gardener/gardener/pull/2480#discussion_r443415041

File: pkg/apis/extensions/v1alpha1/types_worker.go
Comment (jannickfahlbusch) [reply pkg/apis/extensions/v1alpha1/types_worker.go]: Removed due to the comment from @tim-ebert below: https://github.com/gardener/gardener/pull/2480#discussion_r443415041

**Comment by jannickfahlbusch** [comment]
> You can probably add it to:
>
> * nginx-ingress-controller
> * kubernetes-dashboard

I'd like to be able to do this. However I wouldn't consider the plugins as system components. 

How do you see that?

File: example/90-shoot.yaml
Comment (jannickfahlbusch) [reply example/90-shoot.yaml]: Done

File: pkg/apis/core/validation/shoot.go
Comment (jannickfahlbusch) [reply pkg/apis/core/validation/shoot.go]: Done

File: pkg/operation/botanist/worker.go
Comment (jannickfahlbusch) [reply pkg/operation/botanist/worker.go]: Done

File: charts/shoot-core/components/charts/coredns/templates/coredns-deployment.yaml
Comment (jannickfahlbusch) [reply charts/shoot-core/components/charts/coredns/templates/coredns-deployment.yaml]: Done

File: pkg/apis/core/validation/shoot.go
Comment (jannickfahlbusch) [reply pkg/apis/core/validation/shoot.go]: Done

File: pkg/operation/botanist/worker.go
Comment (jannickfahlbusch) [reply pkg/operation/botanist/worker.go]: Removed

File: pkg/apis/core/v1alpha1/defaults_test.go
Comment (jannickfahlbusch) [reply pkg/apis/core/v1alpha1/defaults_test.go]: Changed

**Comment by timebertt** [comment]
> > You can probably add it to:
> > 
> > * nginx-ingress-controller
> > * kubernetes-dashboard
> 
> I'd like to be able to do this. However I wouldn't consider the plugins as system components.
> 
> How do you see that?

Yes, I somewhat agree, but I think, it would be fine to apply the same mechanisms to the addons (if we want via a separate label/field in the shoot spec), but I don't have a strong opinion here.
@gardener/gardener-maintainers any opinions?

File: pkg/apis/core/types_shoot.go
Comment (timebertt) [inline pkg/apis/core/types_shoot.go]:
```suggestion
// WorkerSystemComponents contains configuration for system components related to this worker pool
```

**Comment by rfranzke** [comment]
For the sake of simplicity, I would consider them as "system components" as well (although, they aren't, of course). The rationale behind is mainly that our path forward with the addons is not entirely clear, and adding a new field to the API now that we have to deprecate in the future again is unnecessary work.

**Comment by jannickfahlbusch** [comment]
Sounds good, I'll add the node selectors there as well :)

File: pkg/apis/core/v1alpha1/types_shoot.go
Comment (timebertt) [inline pkg/apis/core/v1alpha1/types_shoot.go]:
```suggestion
// WorkerSystemComponents contains configuration for system components related to this worker pool
```

**Review by rfranzke — COMMENTED** [review]
Nice PR! :) No further comments from my side as of now...
Could you squash the commits?

File: pkg/apis/core/v1beta1/types_shoot.go
Comment (timebertt) [inline pkg/apis/core/v1beta1/types_shoot.go]:
```suggestion
// WorkerSystemComponents contains configuration for system components related to this worker pool
```

File: pkg/apis/core/v1alpha1/defaults.go
Comment (timebertt) [inline pkg/apis/core/v1alpha1/defaults.go]: Hmm, I'm wondering if we should actually default this value in the Shoot object or rather leave it empty if unset and use the `DefaultWorkerSystemComponentsAllow` constant in the shoot controller as a "behavioural" default (set the node label to `true` if the flag in the ShootSpec is not set).
But I am not sure, if this would still be conformant with k8s API conventions.

File: pkg/apis/core/validation/shoot.go
Comment (timebertt) [inline pkg/apis/core/validation/shoot.go]: Hmmm, I think this will allow Shoots with a worker section like this, although we will not be able to schedule system components on either worker pools:
```
workers:
- name: untainted-but-without-system-components
  systemComponents:
    allow: false
- name: tainted-but-system-components-allowed
  taints:
  - key: Foo
     effect: NoSchedule
```
Correct?
I think, we should check here, if the **current** `worker` has no or only compatible taints `systemComponents.allowed=true`.

Comment (jannickfahlbusch) [reply pkg/apis/core/validation/shoot.go]: Done

File: pkg/apis/core/v1alpha1/defaults.go
Comment (jannickfahlbusch) [reply pkg/apis/core/v1alpha1/defaults.go]: I'm not sure. 

I'd like to not have magic where defaulting is happening but not visible on the Shoot CR.
If you prefer it the other way around I can change the behavior

File: pkg/apis/core/validation/shoot_test.go
Comment (danielfoehrKn) [inline pkg/apis/core/validation/shoot_test.go]:
```suggestion
```

**Review by danielfoehrKn — COMMENTED** [review]
/lgtm

File: pkg/apis/core/validation/shoot_test.go
Comment (jannickfahlbusch) [reply pkg/apis/core/validation/shoot_test.go]: Done

File: pkg/apis/core/v1alpha1/defaults.go
Comment (vpnachev) [reply pkg/apis/core/v1alpha1/defaults.go]: The magic can happen with a wrapping function, something like https://github.com/gardener/gardener/blob/7afb12735b14f49144b3a0558eca2b1ceeaba6b0/pkg/apis/core/v1beta1/helper/helper.go#L592-L595

for example:
```go
func SystemComponentsAllowed(worker *gardencorev1beta1.Worker) bool {
	return worker.SystemComponents == nil || worker.SystemComponents.Allow == nil || *worker.SystemComponents.Allow
}
```

**Review by vpnachev — DISMISSED** [review]
/lgtm

File: pkg/apis/core/v1alpha1/defaults.go
Comment (jannickfahlbusch) [reply pkg/apis/core/v1alpha1/defaults.go]: I've changed the behavior in https://github.com/gardener/gardener/pull/2480/commits/8c17a3de043c1e988f916f4ef0e6b5e4993228af.
Please take a look. If it looks good, I'll squash everything together and submit this for a final review

File: pkg/apis/core/v1alpha1/constants/types_constants.go
Comment (timuthy) [inline pkg/apis/core/v1alpha1/constants/types_constants.go]:
```suggestion
	LabelWorkerPoolSystemComponents = "worker.gardener.cloud/system-components"
```
Please also adjust doc string.

File: pkg/apis/core/v1beta1/constants/types_constants.go
Comment (timuthy) [inline pkg/apis/core/v1beta1/constants/types_constants.go]:
```suggestion
	LabelWorkerPoolSystemComponents = "worker.gardener.cloud/system-components"
```
Please also adjust doc string.

File: pkg/apis/core/helper/helpers.go
Comment (rfranzke) [inline pkg/apis/core/helper/helpers.go]:
```suggestion
// SystemComponentsAllowed checks if the given worker allows system components to be scheduled onto it
```

File: pkg/apis/core/types_shoot.go
Comment (rfranzke) [inline pkg/apis/core/types_shoot.go]: Why is this a pointer/an optional field? `WorkerSystemComponents` itself is already optional, so the `Allow` could be a regular field inside it. cc @gardener/gardener-maintainers wdyt?

File: pkg/apis/core/v1alpha1/helper/helper.go
Comment (rfranzke) [inline pkg/apis/core/v1alpha1/helper/helper.go]:
```suggestion
// SystemComponentsAllowed checks if the given worker allows system components to be scheduled onto it
```

File: pkg/apis/core/v1beta1/helper/helper.go
Comment (rfranzke) [inline pkg/apis/core/v1beta1/helper/helper.go]:
```suggestion
// SystemComponentsAllowed checks if the given worker allows system components to be scheduled onto it
```

File: pkg/apis/core/v1alpha1/defaults.go
Comment (rfranzke) [reply pkg/apis/core/v1alpha1/defaults.go]: @tim-ebert why do you suggest this? I don't think we have been consistent here in the past, but generally it's better to be explicit if we already have a dedicated field in the API. Yes, we haven't streamlined this yet for other fields, but for new fields we tried to default it, or? What do other folks think? @timuthy @mvladev @ialidzhikov @danielfoehrKn

File: pkg/apis/core/v1beta1/types_shoot.go
Comment (timuthy) [inline pkg/apis/core/v1beta1/types_shoot.go:962]: Still needed?

**Review by timuthy — CHANGES_REQUESTED** [review]
Just a few nits. Looks good otherwise.

File: pkg/apis/core/types_shoot.go
Comment (timuthy) [reply pkg/apis/core/types_shoot.go]: +1
No necessity to use a pointer.

**Comment by timuthy** [comment]
@jannickfahlbusch concerning the defaulting discussion:
We @gardener/gardener-maintainers had an internal discussion and agreed that we'd prefer an explicit behavior, also because the [K8s API conventions encourage defaulting](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#defaulting).
Can you add the field defaulting functionality again, please? Sorry for the back and forth. We'll also document this best practice to make contributors' lives easier in the future.

File: pkg/apis/core/helper/helpers_test.go
Comment (vpnachev) [inline pkg/apis/core/helper/helpers_test.go]: this package should stay in the second group of imports

File: pkg/apis/core/types_shoot.go
Comment (jannickfahlbusch) [reply pkg/apis/core/types_shoot.go]: Removed

File: pkg/apis/core/v1alpha1/defaults.go
Comment (jannickfahlbusch) [reply pkg/apis/core/v1alpha1/defaults.go]: I've readded the defaulting with https://github.com/gardener/gardener/pull/2480/commits/c159d9da2f77cbaf1a321b71eeeb546a5f76730d but left the helper functions in as I made use of them in the botanist

**Comment by jannickfahlbusch** [comment]
Resolved all requested changes and squashed the commits together.

Please take a look again :)

**Review by rfranzke — DISMISSED** [review]
/lgtm
Thank you very much!
I think it makes sense to add to the release note that system components deployed by extensions (e.g., calico deploys typha and other components) need a separate adaptation and are not covered by this PR.

**Comment by jannickfahlbusch** [comment]
I've adapted the release note.

Do you have a list of components that would need adoption to the feature?
So far I'm only aware of [gardener-extension-networking-calico](https://github.com/gardener/gardener-extension-networking-calico)

**Comment by rfranzke** [comment]
Thanks @jannickfahlbusch, great. If I remember correctly then the calico extension is the only one that deploys non-`DaemonSet` components into the shoot. @zanetworker @DockToFuture what about cilium?

Now that we're talking about it: It probably still makes sense to enhance the https://github.com/gardener/gardener/blob/master/docs/extensions/conventions.md#how-to-create-resources-in-the-shoot-cluster documentation so that extension developers are aware of the fact that they should respect the node labels when deploying components into the shoot.

**Comment by jannickfahlbusch** [comment]
I've added a sentence for the extension documentation and also opened a change for calico: https://github.com/gardener/gardener-extension-networking-calico/pull/39

**Comment by zanetworker** [comment]
@rfranzke Cilium does as well namely: 
- cilium-operator
- hubble

**Review by rfranzke — DISMISSED** [review]

**Review by timuthy — DISMISSED** [review]
/lgtm

**Review by vpnachev — DISMISSED** [review]
/lgtm

**Comment by jannickfahlbusch** [comment]
I had to push again as I forgot to format a test

**Comment by vpnachev** [comment]
The CI validation is failing with:
```
> Check
Executing golangci-lint
Executing go vet
Executing gofmt
Unformatted files detected:
pkg/apis/core/validation/shoot_test.go
```

@jannickfahlbusch  You could run `make check` locally to verify the files are will formatted. 

PS. I was too slow :smile:

**Review by vpnachev — APPROVED** [review]
/lgtm

**Comment by vpnachev** [comment]
/hold

There are some issues and resource manager cannot apply the deployments
```
"Errors occurred during applying: [error during apply of object \"apps/v1/Deployment/kube-system/vpn-shoot\": unable to convert unstructured object to apps/v1, Kind=Deployment: unrecognized type: string error during apply of object \"apps/v1/Deployment/kube-system/coredns\": unable to convert unstructured object to apps/v1, Kind=Deployment: unrecognized type: string error during apply of object \"apps/v1/Deployment/kube-system/metrics-server\": unable to convert unstructured object to apps/v1, Kind=Deployment: unrecognized type: string]",
```

**Comment by jannickfahlbusch** [comment]
> resource manager cannot apply the deployments

How can I debug this to submit a fix?

Edit: I suspect that setting `worker.gardener.cloud/system-components: true` is an issue as `true` evaluates to a Boolean true in YAML.

File: charts/shoot-addons/charts/kubernetes-dashboard/templates/deployment-dashboard.yaml
Comment (vpnachev) [inline charts/shoot-addons/charts/kubernetes-dashboard/templates/deployment-dashboard.yaml]: The value of the selector must be escaped 
```suggestion
        worker.gardener.cloud/system-components: "true"
```

If you try to apply this manifest, on minukube for example or any other k8s cluster,
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: vpn-shoot
  name: vpn-shoot
  namespace: kube-system
spec:
  replicas: 1
  revisionHistoryLimit: 0
  selector:
    matchLabels:
      app: vpn-shoot
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      labels:
        app: vpn-shoot
    spec:
      containers:
      - image: eu.gcr.io/gardener-project/gardener/vpn-shoot:0.19.0
        imagePullPolicy: IfNotPresent
        name: vpn-shoot
      nodeSelector:
        worker.gardener.cloud/system-components: true
      restartPolicy: Always

```

you will get this error
```
Error from server (BadRequest): error when creating "vpn.yaml": Deployment in version "v1" cannot be handled as a Deployment: v1.Deployment.Spec: v1.DeploymentSpec.Template: v1.PodTemplateSpec.Spec: v1.PodSpec.NodeSelector: ReadString: expects " or n, but found t, error found in #10 byte of ...|ponents":true},"rest|..., bigger context ...|ctor":{"worker.gardener.cloud/system-components":true},"restartPolicy":"Always"}}}}

```

**Review by vpnachev — CHANGES_REQUESTED** [review]
The value of the label selector needs to be escaped.

File: charts/shoot-addons/charts/kubernetes-dashboard/templates/deployment-dashboard.yaml
Comment (jannickfahlbusch) [reply charts/shoot-addons/charts/kubernetes-dashboard/templates/deployment-dashboard.yaml]: Fixed

**Review by vpnachev — APPROVED** [review]
/lgtm
/reviewed ok-to-test

**Review by rfranzke — APPROVED** [review]

**Review by timebertt — APPROVED** [review]
Hmm, I was too late... 😄 
But nevertheless, PR looks great!
Thanks a lot for your efforts!

**Pull request merged**
